### PR TITLE
Fix inheriting annotations in dataclasses

### PR DESCRIPTION
**Pull request by sydney-runkle**
## Change Summary

We recently modified how we handle making pydantic `Field`s compatible with `dataclass` fields (various changes in 2.6). This fixes an inheritance issue introduced by those changes.

Thanks @alexmojaki for the starter code for the fix!

## Related issue number

Fix https://github.com/pydantic/pydantic/issues/8670

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @lig

### CODE REVIEW

**Comment by sydney-runkle** [comment]
Please review

File: pydantic/dataclasses.py [hunk 1]
```diff
@@ -153,26 +153,33 @@ def make_pydantic_fields_compatible(cls: type[Any]) -> None:
...
-            # Handle `kw_only` for Python 3.10+
-            if sys.version_info >= (3, 10) and field_value.kw_only:
-                field_args['kw_only'] = True
-
-            # Set `repr` attribute if it's explicitly specified to be not `True`.
-            if field_value.repr is not True:
-                field_args['repr'] = field_value.repr
-
-            setattr(cls, field_name, dataclasses.field(**field_args))
+        for annotation_cls in cls.__mro__[::-1]:
```
Comment (alexmojaki) [inline pydantic/dataclasses.py:156]: why reversed? that seems wrong

**Review by alexmojaki — CHANGES_REQUESTED** [review]
Please also test using the same field name with different type annotations - I think reversing the `__mro__` will break that test.

File: tests/test_dataclasses.py [hunk 2]
```diff
@@ -2759,3 +2759,88 @@ def test_disallow_init_false_and_init_var_true() -> None:
...
+        a: str
+        c: str
+        b: str = pydantic.dataclasses.Field()
+        d: str = pydantic.dataclasses.Field()
+
+    @pydantic.dataclasses.dataclass()
+    class C(B):
+        ...
+
+    c = C(a='a', b='b', c='c', d='d')
```
Comment (alexmojaki) [inline tests/test_dataclasses.py:2843]: Now that all classes are decorated, I think you can go back to checking both B and C in a loop, and then the previous test can be removed.

**Review by alexmojaki — APPROVED** [review]
Thanks!

**Pull request merged**